File: python/p0242_valid_anagram/solutions/hashmap_frequency_balance.py
# Time complexity: O(n + k), practically O(n) if k is small
# Space complexity: O(k)
# where k = number of unique characters


class Solution:
    def isAnagram(self, s: str, t: str) -> bool:

        if len(s) != len(t):
            return False

        freq_map = {}

        for i in range(len(s)):
            freq_map[s[i]] = freq_map.get(s[i], 0) + 1
            freq_map[t[i]] = freq_map.get(t[i], 0) - 1

        # For an anagram, the balance of frequencies is 0 for each character
        return all(count == 0 for count in freq_map.values())

File: rust/p0242_valid_anagram/src/hashmap_frequency_balance.rs
// Time complexity: O(n + k), practically O(n) if k is small
// Space complexity: O(k)
// where k = number of unique characters

use std::collections::HashMap;

pub struct Solution;

impl Solution {
    pub fn is_anagram(s: String, t: String) -> bool {
        if s.len() != t.len() {
            return false;
        }

        let mut freq_map: HashMap<char, i32> = HashMap::new();

        for (s_char, t_char) in s.chars().zip(t.chars()) {
            *freq_map.entry(s_char).or_insert(0) += 1;
            *freq_map.entry(t_char).or_insert(0) -= 1;
        }

        // For an anagram, the balance of frequencies is 0 for each character
        for &freq in freq_map.values() {
            if freq != 0 {
                return false;
            }
        }

        true
    }
}
